### PR TITLE
Support BAM realignment (resolves #17)

### LINKED ISSUE #17: Support BAM realignment in bwakit function

### PR DESCRIPTION
**Pull request by jpfeil**
resolves #17

### CODE REVIEW

File: src/toil_lib/tools/aligners.py [hunk 1]
```diff
@@ -76,6 +76,7 @@ def run_bwakit(job, config, sort=True, trim=False):
         The config must have:
         config.r1               FileStoreID for 1st fastq file
         config.r2               FileStoreID for 2nd fastq file (or None if single-ended)
+        config.bam              FileStoreID for BAM file (realigns BAM or None if aligning fastq)
```
Comment (fnothaft) [inline src/toil_lib/tools/aligners.py]: I would reword as:

> FileStoreID for BAM file to be realigned, or None if aligning fastq

and would update the docs for `config.r[12]`.

File: src/toil_lib/tools/aligners.py [hunk 2]
```diff
@@ -103,47 +104,58 @@ def run_bwakit(job, config, sort=True, trim=False):
...
         rg = config.rg_line
     # Otherwise, generate a read group line to place in the BAM.
-    else:
+    elif all(getattr(config, elem, None) for elem in ['library', 'platform', 'program_unit', 'uuid']):
         rg = "@RG\\tID:{0}".format(config.uuid)  # '\' character is escaped so bwakit gets passed '\t' properly
         rg_attributes = [config.library, config.platform, config.program_unit, config.uuid]
         for tag, info in zip(['LB', 'PL', 'PU', 'SM'], rg_attributes):
             rg += '\\t{0}:{1}'.format(tag, info)
+    else:
+        rg = None
```
Comment (fnothaft) [inline src/toil_lib/tools/aligners.py]: IMO, there's no case where we should allow an unset `@RG` line, because it breaks tons of downstream stuff. Is this change meant to allow `bwakit` to get the `@RG` line(s) from a BAM file header? If so, this is OK, but this fork of the code should require that a BAM file is provided.

**Comment by fnothaft** [comment]
LGTM other than a nit around `@RG` line validation. IIRC, bwakit can realign SAM and CRAM in addition to BAM. Do you know if it does format detection from the input extension or from the file header. Ideally, this patch would support all three, but I wouldn't spend many cycles on it if it is difficult.

**Comment by jvivian** [comment]
@fnothaft — Thanks for reviewing this.  

@jpfeil — Can you look into Frank's question? Then I'll take a final pass and merge.

**Comment by jpfeil** [comment]
bwakit uses the file extension to determine the file format. The documentation says it supports SAM and BAM formats, but I couldn't find anything about CRAM. I tried aligning a CRAM file and it complained.

**Comment by jvivian** [comment]
Thanks @jpfeil — Doing a final pass now.

File: src/toil_lib/tools/aligners.py (hunk 1 above)
Comment (jvivian) [inline src/toil_lib/tools/aligners.py]: +1

**Comment by jvivian** [comment]
Thanks @jpfeil ! LG2M
